basic_bootloader/src/bootloader/block_flow/zk_block_flow/post_tx_op_proving.rs

```diff
@@ -0,0 +1,222 @@
+            );
+        }
+
+        use basic_system::system_implementation::system::public_input::ChainStateCommitment;
+
+        // chain state before
+        let chain_state_commitment_before = ChainStateCommitment {
+            state_root: state_commitment.root,
+            next_free_slot: state_commitment.next_free_slot,
+            block_number: metadata.block_number() - 1,
+            last_256_block_hashes_blake: blocks_hasher.finalize().into(),
+            last_block_timestamp,
+        };
+
+        let mut pubdata_hasher = crypto::blake2s::Blake2s256::new();
+        result_keeper.pubdata(current_block_hash.as_u8_ref());
+        pubdata_hasher.update(current_block_hash.as_u8_ref());
+
+        // Storage
+
+        // 0. Flush accounts into storage, report preimages if needed
+        io.flush_caches(result_keeper);
+        io.report_new_preimages(result_keeper);
+
+        // 1. Return uncompressed NET state diffs for sequencer
+        // It benefit from filter being applied early, so for now it's kept using internal structure
+        result_keeper.storage_diffs(io.storage.storage_cache.net_diffs_iter().map(|(k, v)| {
+            let WarmStorageKey { address, key } = k;
+            let value = v.current_value;
+            (address, key, value)
+        }));
+
+        // 2. Commit to/return compressed pubdata
+        io.storage
+            .apply_storage_diffs_pubdata(result_keeper, &mut pubdata_hasher, &mut io.oracle);
+
+        // Logs pubdata
+        let mut l2_to_l1_logs_hasher = crypto::blake2s::Blake2s256::new();
+        let log_record_fn = |log_hash: &Bytes32| {
+            l2_to_l1_logs_hasher.update(log_hash.as_u8_ref());
+        };
+        // use concrete type as it's non-trivial
+        io.logs_storage.apply_logs_to_pubdata_and_record_log_hashes(
+            result_keeper,
+            &mut pubdata_hasher,
+            Some(log_record_fn),
+        );
+        // Logs themselves
+        result_keeper.logs(io.logs_storage.messages_ref_iter());
+
+        // Events - no pubdata
+        result_keeper.events(io.events_storage.events_ref_iter());
+
+        // 3. Verify/apply reads and writes
+        cycle_marker::wrap!("verify_and_apply_batch", {
+            IOTeardown::<_>::update_commitment(
+                &mut io,
+                Some(&mut state_commitment),
+                &mut logger,
+                result_keeper,
+            );
+        });
+
+        let pubdata_hash = pubdata_hasher.finalize();
+        let l2_to_l1_logs_hashes_hash = l2_to_l1_logs_hasher.finalize();
+
+        let mut blocks_hasher = crypto::blake2s::Blake2s256::new();
+        blocks_hasher.update(current_block_hash.as_u8_ref());
+        for depth in 0..255 {
+            blocks_hasher.update(
+                metadata
+                    .block_historical_hash(depth)
+                    .expect("must be known for such depth")
+                    .as_u8_ref(),
+            );
+        }
```

Comment:
I think this is different to how it's calculated in post_tx_op_batch_proving. My understanding is that this one is correct, so we need to fix the other one, right?

Reply:
This one is consistent with block hashes structure comment (that the most recent is at index 0), so pre-pending is consistent with such comment. Maybe the logic changed for some reason